sdk/servicebus/service-bus/samples/typescript/src/advanced/sessionRoundRobin.ts

```diff
@@ -0,0 +1,146 @@
+
+dotenv.config();
+
+const serviceBusConnectionString =
+  env["SERVICEBUS_CONNECTION_STRING"] || "<service bus connection string not in environment>";
+
+// NOTE: this sample uses a queue but would also work a session enabled subscription.
+const queueName = env["QUEUE_NAME_WITH_SESSIONS"] || "<queue name not in environment>";
+
+const maxSessionsToProcessSimultaneously = 8;
+const sessionIdleTimeoutMs = 3 * 1000;
+const delayWhenNoSessionsAvailableMs = 5 * 1000;
+
+// this can be used control when the round-robin processing will terminate.
+const abortController = new AbortController();
+
+// called just before we start processing the first message of a session
+async function processInitialize(sessionId: string) {
+  console.log(`[${sessionId}] will start processing...`);
+}
+
+// called when we get a message for a session
+async function processMessage(msg: ReceivedMessageWithLock) {
+  console.log(`[${msg.sessionId}] received message with body ${msg.body}`);
+}
+
+// called if we get an error
+async function processError(err: Error, sessionId?: string) {
+  console.log(`[${sessionId}] had error ${err.message}`);
+}
+
+// Called if we are closing a session
+// `reason` will be:
+// * 'error' if we are closing because of an error(the error will be delivered
+//   to `processError` above)
+// * 'idle_timeout' if we `sessionIdleTimeoutMs` milliseconds pass without
+//   any messages being received (ie, session can be considered empty).
+//
+async function processClose(reason: "error" | "idle_timeout", sessionId: string) {
+  if (reason === "error") {
+    console.log(`[${sessionId}] was closed because of error`);
+  } else if (reason === "idle_timeout") {
+    console.log(`[${sessionId}] was closed - no more messages within idle timeout`);
+  }
+}
+
+// utility function to create a timer that can be refreshed
+function createRefreshableTimer(timeoutMs: number, resolve: Function): () => void {
+  let timer: any;
+
+  return () => {
+    clearTimeout(timer);
+    timer = setTimeout(() => resolve(), timeoutMs);
+  };
+}
+
+// Queries Service Bus for the next available session and processes it.
+async function receiveFromNextSession(serviceBusClient: ServiceBusClient): Promise<void> {
+  let sessionReceiver: SessionReceiver<ReceivedMessageWithLock>;
+
+  try {
+    sessionReceiver = await serviceBusClient.createSessionReceiver(queueName, "peekLock", {
+      autoRenewLockDurationInMs: sessionIdleTimeoutMs
+    });
+  } catch (err) {
+    if ((err as MessagingError).code === "SessionCannotBeLockedError") {
+      console.log(`INFO: no available sessions, sleeping for ${delayWhenNoSessionsAvailableMs}`);
+      await delay(delayWhenNoSessionsAvailableMs);
+      return;
+    }
+
+    await processError(err, undefined);
+    return;
+  }
+
+  await processInitialize(sessionReceiver.sessionId);
+
+  const sessionFullyRead = new Promise((resolveSessionAsFullyRead, rejectSessionWithError) => {
```

Comment:
When the timer runs out, the promise `sessionFullyRead` is resolved.
When there is an error, the promise `sessionFullRead` is regjected

But, neither ensures that we stop receiving messages right away.
In this sample, we await on `processClose()` and then close the session receiver.

Am guessing this is because we currently don't have a way to stop receiving messages without closing the message receiver? If so, then until we have this feature, can we pass an abort signal to `subscribe()` instead?

Reply:
We do pass an abortSignal to subscribe?

Reply:
I meant that the subscribe method is not stopped right away after the promise is resolved/rejected
If the subscribe method returned a closable thing, then we would call it before calling `processClose`.

The abort signal being passed is the uber signal that is passed to the main function of this sample. I was referring to an abort signal that would get fired for when `sessionFullyRead` is resolved or rejected. You can use link the two signals such that subscribe is aborted when either is fired

Reply:
And in both cases the receiver is probably about to exit w/o any more messages _anyways_:
* The session is probably empty and isn't getting any more messages (the reason the idle timer fired in the first place).
* An error happened in which case rhea is about to close the link anyways.

We could add more error handling code in here but for sample purposes I think it's covered well enough.

Reply:
Fair point on the second point, but there is a corner case of users `processClose()` taking long enough for a message to come by in case of idle timeout :)

But I agree, for the sake of this sample, we are good.